gammagl/models/specformer.py

```diff
@@ -0,0 +1,298 @@
+        :return:
+        """
+        # default:self attention
+        keys = querys
+        values = querys
+
+        d = tlx.get_tensor_shape(querys)[-1]
+        scores = tlx.ops.bmm(querys, tlx.nn.Transpose(perm=[0, 2, 1])(keys)) / math.sqrt(d)
+        self.attn_weights = tlx.nn.Softmax(axis=-1)(scores)
+        return tlx.ops.bmm(self.attn_weights, values)
+class MultiHeadAttention_tlx(tlx.nn.Module):
+    def __init__(self, hidden_dim, nheads, tran_dropout=0.0):
+        super().__init__(self)
+        self.num_heads = nheads
+        self.attention = DotProductAttention_tlx()
+
+        self.W_q = tlx.nn.Linear(in_features=hidden_dim, out_features=hidden_dim, act=tlx.ReLU)
+        self.W_k = tlx.nn.Linear(in_features=hidden_dim, out_features=hidden_dim, act=tlx.ReLU)
+        self.W_v = tlx.nn.Linear(in_features=hidden_dim, out_features=hidden_dim, act=tlx.ReLU)
+        self.W_o = tlx.nn.Linear(in_features=hidden_dim, out_features=hidden_dim, act=tlx.ReLU)
+
+    def forward(self, q, k, v):
+        is_batched = len(tlx.get_tensor_shape(q)) == 3
+
+        if not is_batched:  # if input with no batchs
+            q = tlx.ops.expand_dims(q, axis=0)
+            k = tlx.ops.expand_dims(k, axis=0)
+            v = tlx.ops.expand_dims(v, axis=0)
+
+        q = transpose_qkv_tlx(self.W_q(q), self.num_heads)
+        k = transpose_qkv_tlx(self.W_k(k), self.num_heads)
+        v = transpose_qkv_tlx(self.W_v(v), self.num_heads)
+
+        output = self.attention(q, k, v)
+        output_concat = transepose_output_tlx(output, self.num_heads)
+        res = self.W_o(output_concat)
+
+        if not is_batched:  # if input with no batchs,remove the batchs which is added in the begining
+            res = tlx.ops.squeeze(res, axis=0)
+
+        return res
+class SineEncoding(tlx.nn.Module):
+    def __init__(self, hidden_dim=16):
+        super(SineEncoding, self).__init__()
+        self.constant = 100
+        self.hidden_dim = hidden_dim
+        initor = tlx.nn.initializers.XavierNormal()
+        self.eig_w = tlx.nn.Linear(in_features=hidden_dim + 1, out_features=hidden_dim,  # h+1  -> h
+                                   act=tlx.ReLU,
+                                   W_init=initor)
+
+    def forward(self, e):
+        # input:  [N]
+        # output: [N, d]
+        ee = e * self.constant
+        div = tlx.ops.exp(tlx.ops.arange(0, self.hidden_dim, 2, dtype=tlx.float32) * (-math.log(10000) /
+                                                                                      self.hidden_dim))
+        pe = tlx.ops.expand_dims(ee, axis=1) * div
+        eeig = tlx.ops.concat((tlx.ops.expand_dims(e, axis=1), tlx.ops.sin(pe), tlx.ops.cos(pe)),
+                              axis=1)
+        eeig = tlx.cast(eeig, dtype=tlx.float32)
+        # print(eeig)
+        return self.eig_w(eeig)
+class FeedForwardNetwork_tlx(tlx.nn.Module):

```

Comment:
What is this doing? Why not directly use a two-layer linear layer in the following class?